stan/math/prim/prob/exp_mod_normal_cdf.hpp

```diff
@@ -24,104 +27,118 @@ return_type_t<T_y, T_loc, T_scale, T_inv_scale> exp_mod_normal_cdf(
+  ref_type_t<decltype(value_of(lambda_arr))> lambda_val = value_of(lambda_arr);
 
-  for (size_t n = 0; n < size_y; n++) {
-    if (is_inf(y_vec[n]) && y_vec[n] < 0.0) {
-      return ops_partials.build(0.0);
-    }
-  }
+  check_not_nan(function, "Random variable", y_val);
+  check_finite(function, "Location parameter", mu_val);
+  check_positive_finite(function, "Scale parameter", sigma_val);
+  check_positive_finite(function, "Inv_scale parameter", lambda_val);
 
-  for (size_t n = 0; n < N; n++) {
-    const T_partials_return y_dbl = value_of(y_vec[n]);
-    const T_partials_return mu_dbl = value_of(mu_vec[n]);
-    const T_partials_return sigma_dbl = value_of(sigma_vec[n]);
-    const T_partials_return lambda_dbl = value_of(lambda_vec[n]);
-    const T_partials_return inv_sigma = inv(sigma_dbl);
-    const T_partials_return diff = y_dbl - mu_dbl;
-    const T_partials_return u = lambda_dbl * diff;
-    const T_partials_return v = lambda_dbl * sigma_dbl;
-    const T_partials_return scaled_diff = diff * INV_SQRT_TWO * inv_sigma;
-    const T_partials_return scaled_diff_diff = scaled_diff - v * INV_SQRT_TWO;
-    const T_partials_return erf_calc = 0.5 * (1 + erf(scaled_diff_diff));
-    const T_partials_return exp_term = exp(0.5 * square(v) - u);
-    const T_partials_return exp_term_2 = exp(-square(scaled_diff_diff));
+  if (size_zero(y, mu, sigma, lambda)) {
+    return 1.0;
+  }
 
-    const T_partials_return deriv_1 = lambda_dbl * exp_term * erf_calc;
-    const T_partials_return deriv_2
-        = INV_SQRT_TWO_PI * exp_term * exp_term_2 * inv_sigma;
-    const T_partials_return deriv_3
-        = INV_SQRT_TWO_PI * exp(-square(scaled_diff)) * inv_sigma;
+  operands_and_partials<T_y_ref, T_mu_ref, T_sigma_ref, T_lambda_ref>
+      ops_partials(y_ref, mu_ref, sigma_ref, lambda_ref);
+  if (sum(y_val == NEGATIVE_INFTY)) {  // here sum works as "any"
+    return ops_partials.build(0.0);
+  }
 
-    const T_partials_return cdf_n
-        = 0.5 + 0.5 * erf(scaled_diff) - exp_term * erf_calc;
+  const auto& inv_sigma
+      = to_ref_if<!is_constant_all<T_y, T_loc, T_scale>::value>(inv(sigma_val));
+  const auto& diff = to_ref(y_val - mu_val);
+  const auto& v = to_ref(lambda_val * sigma_val);
+  const auto& scaled_diff = to_ref(diff * INV_SQRT_TWO * inv_sigma);
+  const auto& scaled_diff_diff
+      = to_ref_if<!is_constant_all<T_y, T_loc, T_scale, T_inv_scale>::value>(
+          scaled_diff - v * INV_SQRT_TWO);
+  const auto& erf_calc = to_ref(0.5 * (1 + erf(scaled_diff_diff)));
 
-    cdf *= cdf_n;
+  // we can not have this call as a rvalue in next expression until square is
+  // returning expressions.
+  const auto& v2 = square(v);
+  const auto& exp_term
+      = to_ref_if<!is_constant_all<T_y, T_loc, T_scale, T_inv_scale>::value>(
+          exp(0.5 * v2 - lambda_val * diff));
+  const auto& cdf_n
+      = to_ref(0.5 + 0.5 * erf(scaled_diff) - exp_term * erf_calc);
```

Comment:
Related question, but is this okay: https://github.com/stan-dev/math/pull/2023/files#diff-ea1615a62ecafd356acb8462ebcb731fR95 ? If so what is the difference?

```cpp
const auto& exp_m_sq_inner_term = exp(-square(inner_term));
```

And then this variation: https://github.com/stan-dev/math/pull/2023/files#diff-d52cfdf251ec9944cc3d32637b9dff92R102

```cpp
const auto& exp_term_2 = to_ref_if<(!is_constant_all<T_y, T_loc, T_scale>::value &&
  !is_constant_all<T_inv_scale>::value)>(exp(-square(scaled_diff_diff)));
```

Reply:
Yep, right.

The difference is that the last/top function in the expression is also from stan math. When it evaluates the expression it creates a temporyry matrix. However in this case the temporary is assigned to a const&, extending its lifetime to that of the reference. This will also be fine once all these functions start returning expressions, assuming that happens for all at the same time.

In the last example `to_ref_if` is unnecessary at the moment, but it is here just so we don't have to change the code once functions start returning expressions.

Reply:
> However in this case the temporary is assigned to a const&

Oh wow.

Why does something like:

```
0.5 * square(v) - lambda_val * diff
```

turn into an expression that might lose the memory allocated in square and then

```
-square(v)
```

doesn't? Is the negation operator there some lighter weight?

Reply:
Nope. All functions are equal (Not true actually ... If we do `using std::square;`, Eigen overload is chosen, which returns expression. But that is not what is happening here). What is happening is that `exp` is evaluated on this line, so it does not matter if its argument references any temporaries.

Reply:
> What is happening is that exp is evaluated on this line

Apologies for not getting this, but it's still not clear to me.

If these two lines evaluate the exp and don't care about temporaries:

A - `const auto& exp_m_sq_inner_term = exp(-square(inner_term));`

B - `const auto& exp_term_2 = to_ref_if<(!is_constant_all<T_y, T_loc, T_scale>::value &&
  !is_constant_all<T_inv_scale>::value)>(exp(-square(scaled_diff_diff)));`

Then what's the difference with this that makes temporaries?

C - `const auto& exp_term
      = to_ref_if<!is_constant_all<T_y, T_loc, T_scale, T_inv_scale>::value>(
          exp(0.5 * v2 - lambda_val * diff));`

From you response earlier:

> The difference is that the last/top function in the expression is also from stan math.

I don't see how B calls a different exp. Is it std::square or std::exp in some cases and Stan in another? I can hangouts tomorrow if it's annoying to keep trying to type these things out to me and wait 24 hours to realize I'm still confused. Is 8:30AM Pacific too late in your day?

I'll try to review your distribution pulls tomorrow to get them movin' too.

Reply:
The problem happens only if we create an expression referencing temporaries and don not evaluate it on the same line. If a temporary is assigned to a reference its lifetime is extended to that of the reference and it does not go out of scope before expression is evaluated. If the expression referencing temporaries is evaluated on the same line as temporaries are created, it is also fine. In this case the expression is evaluated before the temporaries get destroyed.

If that is still not clear I can make it on 8:30 pacific.